### PR TITLE
SHM part 2

### PR DESCRIPTION
**Pull request by NikczemnyOkularnik**
@Kasia9311  @kondorski  @Crazyayane07  @pwiatr3 @NikczemnyOkularnik

### CODE REVIEW

File: shm/Game.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+class Map;
+
+Game::Game(size_t money, size_t gameDays, size_t finalGoal)
+    : time_(new Time())
+    , money_(money)
+    , gameDays_(gameDays)
+    , finalGoal_(finalGoal)
+    , menu_(std::make_unique<Menu>(this))
+    , map_(new Map(time_))
+    , playerOne_(new Player(std::make_unique<Ship>(20, 30, 10, "Dar Pomorza", 3, time_), 1000, 1000))
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp]: I think all should be constexpr variables

File: shm/Game.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+{
+    printTitle();
+    setPlayer();  
+    do{
+        if (playerOne_->getMoney() >= 2000)
+        {
+            endGameWin();
+            quitRequested();
+        }
+        if (playerOne_->getMoney() < 0 || time_->getCurrentTime() >= gameDays_)
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp]: else if

File: shm/Game.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        {
+            endGameLose();
+            quitRequested();
+        }
+        if (quitRequest)
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:41]: else if

File: shm/Game.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        {
+            break;
+        }
+        
+        menu_->printMenu();
+        menu_->playerChoice();
+
+    }while(!quitRequest);
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp]: you always quit earlier in if (quitRequest). So here it is redundant.

File: shm/Game.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    map.PrintCurrentPosition();
+    map.DebugPrintIsland();
+}
+
+void Game::travel()
+{
+    auto i = 0;
+    printMap(*map_);
+    std::cout << "Choose Your destination captain!" << '\n';
+    std::cin >> i;
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:87]: What when someone provides the wrong input? Like "Ala ma kota"

File: shm/Game.cpp [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+{
+    playerOne_->getShip()->printShipCargo();
+}
+
+void Game::setStartingCargo()
+{
+    playerOne_->getShip()->shipCargo.push_back(new Fruit("Banany", 1, 40, time_, 5, 0)); 
+    playerOne_->getShip()->shipCargo.push_back(new Fruit("Apple", 1, 14, time_, 20, 0));
+    playerOne_->getShip()->shipCargo.push_back(new Alcohol("Rum", 1, 60, time_, 70));        
+    playerOne_->getShip()->shipCargo.push_back(new Item("Hook", 1, 100, time_, Rarity::common));
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp]: "new" is depreciated. Please use make_shared/ make_unique.

File: shm/Game.cpp [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    std::cout << "This store contains: " << '\n';
+    currentStore_->printStoreCargo();
+}
+
+void Game::buyCargo()
+{
+    int cargoElement = 0;
+    size_t amount = 0;
+    std::cout << "Choose cargo: ";
+    std::cin >> cargoElement;
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:145]: validate input

File: shm/Game.cpp [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    
+}
+
+void Game::sellCargo()
+{
+    int cargoElement = 0;
+    size_t amount = 0;
+    std::cout << "Choose cargo: ";
+    std::cin >> cargoElement;
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:166]: same here

File: shm/Map.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    islands_.reserve(constVariables::ISLANDS_COUNT);
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_real_distribution<> distrib(constVariables::COORDINATE_MIN , constVariables::COORDINATE_MAX);
+
+    std::vector<Coordinates> cords;
+    cords.reserve(constVariables::ISLANDS_COUNT);
+    for (int i = 0; i < constVariables::ISLANDS_COUNT; i++)
+    {   
+        Coordinates c ((float)distrib(gen), (float)distrib(gen));
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp]: don't use "C style" cast, use static_cast instead

File: shm/Map.cpp [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    return std::find(vec.begin(), vec.end(), c) != vec.end();
+}
+
+size_t Map::calculateDistance(Island island_pos_)
+{
+    auto distanceX = current_pos_->getPosition().GetPositionX();
+    auto distanceY = current_pos_->getPosition().GetPositionY();
+    auto disX = island_pos_.getPosition().GetPositionX();
+    auto disY = island_pos_.getPosition().GetPositionY();
+    auto distance = sqrt(pow((disX-distanceX), 2) + pow((disY - distanceY),2));
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp]: make it const

File: shm/Map.cpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+void Map::PrintCurrentPosition()
+{
+     std::cout <<"Current coordinates: " << std::to_string(current_pos_->getPosition().GetPositionX()) << " | " << std::to_string(current_pos_->getPosition().GetPositionY()) << '\n';
+}
+
+bool Map::contains(const std::vector<Coordinates>& vec, const Coordinates& c) {
+    return std::find(vec.begin(), vec.end(), c) != vec.end();
+}
+
+size_t Map::calculateDistance(Island island_pos_)
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:72]: seems like const method

File: shm/Map.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        j++;
+    }
+}
+
+void Map::PrintCurrentPosition()
+{
+     std::cout <<"Current coordinates: " << std::to_string(current_pos_->getPosition().GetPositionX()) << " | " << std::to_string(current_pos_->getPosition().GetPositionY()) << '\n';
+}
+
+bool Map::contains(const std::vector<Coordinates>& vec, const Coordinates& c) {
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp]: const method

File: shm/Map.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+void Map::DebugPrintIsland () {
+    int j = 0;
+    for (auto &el : islands_) {
+        std::cout << j << " | "<< std::to_string(el.getPosition().GetPositionX()) << " | " << std::to_string(el.getPosition().GetPositionY()) << '\n';
+        j++;
+    }
+}
+
+void Map::PrintCurrentPosition()
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:62]: cont method

File: shm/Map.cpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            c.SetPositionX(distrib(gen));
+            c.SetPositionY(distrib(gen));
+        }
+        islands_.push_back(Island (c, time));
+        cords.push_back(c);
+    } 
+}
+
+void Map::DebugPrintIsland () {
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp]: cont method

File: shm/Map.hpp [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include "island.hpp"
+#include <vector>
+#include "Time.hpp"
+
+//class Island;
```
Comment (nauka-programowania-MA) [inline shm/Map.hpp:6]: ?

File: shm/Map.hpp [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+class Coordinates;
+
+//Class responsible for representing map in game
+class Map {
+public:
+    Map(Time * time);
+    void DebugPrintIsland();
+    void PrintCurrentPosition();
+    friend class Game;
+    Island* getCurrentPos() {return current_pos_;}
```
Comment (nauka-programowania-MA) [inline shm/Map.hpp:16]: cont method

File: shm/Map.hpp [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+//class Island;
+class Coordinates;
+
+//Class responsible for representing map in game
+class Map {
+public:
+    Map(Time * time);
+    void DebugPrintIsland();
+    void PrintCurrentPosition();
+    friend class Game;
```
Comment (nauka-programowania-MA) [inline shm/Map.hpp:15]: don't mix function with friend declaration

File: shm/Menu.cpp [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    menuChoice(item_);
+}
+
+bool Menu::isPlayerChoiceValid(const size_t &playerAnswer)
+{
+    if(std::cin.fail() || playerAnswer < 0 || playerAnswer > 5)
+    {
+        std::cin.clear();
+        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+        std::cout << "Wrong answer, please use numbers 0 - 5" << '\n';
```
Comment (nauka-programowania-MA) [inline shm/Menu.cpp:27]: Nice. Do the same in others input method

File: shm/Menu.hpp [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        Exit,
+    };
+
+    Menu(Game*);
+    ~Menu();
+
+    bool isPlayerChoiceValid(const size_t &playerAnswer);
+    void menuChoice(MenuItem);
+    void printMenu();
+    void playerChoice();
```
Comment (nauka-programowania-MA) [inline shm/Menu.hpp]: I guess most of them could be const

File: shm/Storable.cpp [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+{
+    if(moneyToEarn > 0)
+    {
+        money_ += moneyToEarn;
+    }
+}
+void Storable::removeCargo(std::shared_ptr<Cargo>, size_t)
+{
+    
+}
```
Comment (nauka-programowania-MA) [inline shm/Storable.cpp]: ?

File: shm/Storable.hpp [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+        {}
+
+    //Constructor for objects subscribed to Time class (i.e. Store)
+    Storable(int money, size_t availableSpace, Time* time)
+        : money_(money)
+        , availableSpace_(availableSpace)
+        , time_(time)
+        {
+            time_->attach(this);
+        }
```
Comment (nauka-programowania-MA) [inline shm/Storable.hpp]: move C'tor definition to .cpp

File: shm/Store.cpp [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+#include "Time.hpp"
+#include "Store.hpp"
+
+constexpr size_t numberOfPossibleFruitCargo = 4;
+constexpr size_t numberOfPossibleAlcoholCargo = 4;
+constexpr size_t numberOfPossibleItemCargo = 4;
+std::vector<std::string> possibleFruitCargoNames = {"Banana", "Pineapple", "Apple", "Pear"};
+std::vector<std::string> possibleAlcoholCargoNames = {"Whiskey", "Rum", "Vodka", "Beer"};
+std::vector<std::string> possibleItemCargoNames = {"Rope", "Hook", "Screw", "Spoon"};
+std::vector<Rarity> rarityChoice = {Rarity::common, Rarity::rare, Rarity::epic, Rarity::legendary};
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:14]: make it constexpr array.
Instad strign use array<const* char>

File: shm/Store.cpp [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+    size_t fruitCargoNumber = fruitCargoDistrib(gen);
+    size_t alcoholCargoNumber = alcoholCargoDistrib(gen);
+    size_t itemCargoNumber = itemCargoDistrib(gen);
+
+    size_t fruitPriceX = fruitPrice(gen);
+    size_t alcoholPriceX = alcoholPrice(gen);
+    size_t itemPriceX = itemPrice(gen);
+
+    size_t cargoQuantity = quantity(gen);
+    size_t alcoholPercentage = percentage(gen);
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp]: const

File: shm/Store.cpp [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+
+    std::string fruitCargoName = possibleFruitCargoNames[fruitCargoNumber];
+    std::string alcoholCargoName = possibleAlcoholCargoNames[alcoholCargoNumber];
+    std::string itemCargoName = possibleItemCargoNames[itemCargoNumber];
+
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:60]: const

File: shm/Store.cpp [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+    storeCargo.push_back(new Fruit(fruitCargoName, (cargoQuantity+10), fruitPriceX, time, 15, 0)); 
+    storeCargo.push_back(new Alcohol(alcoholCargoName, cargoQuantity, alcoholPriceX, time, alcoholPercentage)); 
+    storeCargo.push_back(new Item(itemCargoName, (cargoQuantity/2), itemPriceX, time, randomRarity)); 
+
+}
+
+Cargo * Store::findMatchCargo(Cargo * cargo)
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp]: const method

File: shm/Store.cpp [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        return Response::lack_of_cargo;
+    }
+
+    return Response::done;
+}
+
+Cargo * Store::getCargo(size_t index) const
+{
+    return nullptr;
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:165]: ?

File: shm/Store.cpp [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+void Store::nextDay(size_t elapsedTime)
+{
+    for (auto el : storeCargo)
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:170]: here you copy

File: shm/Store.cpp [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        el->reduceAmount(1);
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:172]: and work on copy

File: shm/Store.cpp [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+
+void Store::printStoreCargo()
+{
+    auto i = 0;
+    for (auto el : storeCargo)
+    {
+        std::cout << '\n';
+        el->printCargo(i);
+        i++;
+    }
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:184]: if you need index in loop, use old loop instead. for(int i = 0....)

File: shm/Store.cpp [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+void Store::removeCargo(Cargo * item, size_t amount)
+{
+    auto storeCargoAmount = findMatchCargo(item)->getAmount();
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:189]: const

File: shm/Store.cpp [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+    if(storeCargoAmount == amount)
+    {
+        auto i = std::find(begin(storeCargo), end(storeCargo), item);
+        storeCargo.erase(i);
+    }
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:194]: return here and remove else

File: shm/Store.cpp [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+}
+void Store::addStoreCargo(Cargo * item)
+{
+    auto cargoPtr = findMatchCargo(item);
+    if(findMatchCargo(item))
+    {
+        findMatchCargo(item)->increaseAmount(item->getAmount());
+    }
+    else
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:208]: return form if and remove else

File: shm/Store.cpp [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+        storeCargo.erase(i);
+    }
+    else
+    {
+        findMatchCargo(item)->reduceAmount(amount);
+    }
+}
+void Store::addStoreCargo(Cargo * item)
+{
+    auto cargoPtr = findMatchCargo(item);
```
Comment (nauka-programowania-MA) [inline shm/Store.cpp:203]: const

File: shm/Store.hpp [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+#include "fruit.hpp"
+#include "item.hpp"
+#include "TimeEffectable.hpp"
+#include "cargo.hpp"
+#include <random>
+
+class Time;
+
+class Player;
+//class Cargo;
```
Comment (nauka-programowania-MA) [inline shm/Store.hpp:16]: ?

File: shm/Store.hpp [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+
+enum class Response
```
Comment (nauka-programowania-MA) [inline shm/Store.hpp:18]: Move to class

File: shm/Store.hpp [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+    Response sell(Cargo * cargo, size_t amount, Player* player);
+
+    Cargo * getCargo(size_t) const override;
+    Cargo * findMatchCargo(Cargo *);
+    std::vector<Cargo *> storeCargo;
+
+    void addStoreCargo(Cargo *);
+    void removeCargo(Cargo *, size_t);    
+    void nextDay(size_t) override;
+    void printStoreCargo();
```
Comment (nauka-programowania-MA) [inline shm/Store.hpp:41]: should be const method

File: shm/Store.hpp [hunk 37]
```diff
@@ -0,0 +1,53 @@
...
+
+    friend std::ostream& operator<<(std::ostream& out, const Store& store) {
+        out << "Store: \n";
+        for (auto el : store.storeCargo) {
+            out << el->getCargoInfo();
+        }
+
+        return out;
+    }
```
Comment (nauka-programowania-MA) [inline shm/Store.hpp:50]: move definition to .cpp

File: shm/Time.cpp [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+Time::~Time(){}
+
+void Time::attach(TimeEffectable* observer)
+{   
+    list_observer_.push_back(observer);
+}
+
+void Time::detach(TimeEffectable* observer)
+{
+    list_observer_.remove(observer);
```
Comment (nauka-programowania-MA) [inline shm/Time.cpp:20]: You should attach/ dettach in D'tor. Use RAII pattern

File: shm/Time.cpp [hunk 39]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+void Time::notify(size_t elapsedDays)
+{
+    for(auto el : list_observer_)
+    {
+        el->nextDay(elapsedDays);
+    }
```
Comment (nauka-programowania-MA) [inline shm/Time.cpp]: Do not copy

File: shm/TimeEffectable.hpp [hunk 40]
```diff
@@ -0,0 +1,6 @@
+#pragma once
+
+class TimeEffectable {
+public:
+    virtual void nextDay(size_t) {}
```
Comment (nauka-programowania-MA) [inline shm/TimeEffectable.hpp:5]: Add virtual D'tor when you have virtual method

File: shm/fruit.cpp [hunk 41]
```diff
@@ -0,0 +1,56 @@
...
+ Fruit::Fruit(const std::string &name, size_t amount, size_t basePrice,  Time *time, size_t expiry_date, size_t time_elapsed)
+        : Cargo(name, amount, basePrice, time)
+        , time_elapsed_(time_elapsed)
+        , expiry_date_(expiry_date)
+        {}
+
+Fruit::~Fruit()
+{
+    // this->time_.detach(this);
+}
```
Comment (nauka-programowania-MA) [inline shm/fruit.cpp]: Yup, attach in detach in C'tor/ D'tor

File: shm/fruit.cpp [hunk 42]
```diff
@@ -0,0 +1,56 @@
...
+
+ Fruit::Fruit(const std::string &name, size_t amount, size_t basePrice,  Time *time, size_t expiry_date, size_t time_elapsed)
+        : Cargo(name, amount, basePrice, time)
+        , time_elapsed_(time_elapsed)
+        , expiry_date_(expiry_date)
+        {}
+
+Fruit::~Fruit()
+{
+    // this->time_.detach(this);
```
Comment (nauka-programowania-MA) [inline shm/fruit.cpp]: You have crash, here because you didn't attach observer in C'tor, you only wanted to remove it here

File: shm/ship.cpp [hunk 43]
```diff
@@ -0,0 +1,179 @@
...
+    : capacity_(capacity), crew_(Crew), speed_(speed), name_(name), id_(id), time_(time)
+{
+    time_->attach(this);
+    shipCargo.reserve(10);
+}
+
+Ship::Ship(int maxCrew, int speed, size_t id, Time *time)
+    : Ship(0, maxCrew, speed, "", id, time)
+{
+    time_->attach(this);
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: This is wrong! You attach twice the same observer!!!

Please attach it in one C'tor, and rest C'tors will call this one

File: shm/ship.cpp [hunk 44]
```diff
@@ -0,0 +1,179 @@
...
+void Ship::addCargo(Cargo * item)
+{
+    auto cargoPtr = findMatchCargo(item);
+    if(findMatchCargo(item))
+    {
+        findMatchCargo(item)->increaseAmount(item->getAmount());
+    }
+    else
+    {
+        shipCargo.emplace_back(item);
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: push_back, use emplace when you pass arguments to C'tro, here you use copy C'tor or Move C'tor.

File: shm/ship.cpp [hunk 45]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+void Ship::removeCargo(Cargo * item, size_t amount)
+{
+    auto shipCargoAmount = findMatchCargo(item)->getAmount();
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: const

File: shm/ship.cpp [hunk 46]
```diff
@@ -0,0 +1,179 @@
...
+    if(shipCargoAmount == amount)
+    {
+        auto it = std::find_if(begin(shipCargo), end(shipCargo), [item](const auto* el){ return *el == *item; });
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: const

File: shm/ship.cpp [hunk 47]
```diff
@@ -0,0 +1,179 @@
...
+        if(Fruit* f = dynamic_cast<Fruit*>(loadCargo)){
+            toAdd = new Fruit(*f);
+
+        }
+        if(Item* i = dynamic_cast<Item*>(loadCargo)){
+            toAdd = new Item(*i);
+        }
+        if(Alcohol* a = dynamic_cast<Alcohol*>(loadCargo)){
+            toAdd = new Alcohol(*a);
+        }
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: new is deprecated. Please create 'clone()` method to cargo, which will clone itself and return unique_ptr to the new object. When you do this, you avoid ugly dynamic_cast which in 95% show the wrong style of code.

`const auto* toAdd = loadCargo->clone()`

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]

File: shm/Game.cpp (hunk 7 above)
Comment (Kasia9311) [reply shm/Game.cpp:145]: PR 6e5d003

**Comment by ziobron** [comment]
Thanks @Kasia9311 for Code Review.
🏅 1 XP granted

File: shm/Game.cpp (hunk 8 above)
Comment (Kasia9311) [reply shm/Game.cpp:166]: PR 6e5d003

**Comment by ziobron** [comment]
Thanks @Kasia9311 for Code Review.
🏅 1 XP granted

File: shm/Game.cpp (hunk 3 above)
Comment (Kasia9311) [reply shm/Game.cpp:41]: PR e150a0e

**Comment by ziobron** [comment]
Thanks @Kasia9311 for Code Review.
🏅 1 XP granted

File: shm/Game.cpp (hunk 2 above)
Comment (Kasia9311) [reply shm/Game.cpp]: PR e150a0e

File: shm/Game.cpp (hunk 4 above)
Comment (Kasia9311) [reply shm/Game.cpp]: PR cf1bed7

File: shm/Game.cpp (hunk 5 above)
Comment (Kasia9311) [reply shm/Game.cpp:87]: PR 681f9f0

File: shm/Map.cpp (hunk 10 above)
Comment (Kasia9311) [reply shm/Map.cpp]: PR 08f0381

**Review by ziobron — COMMENTED** [review]
Just checking automation

**Comment by ziobron** [comment]
Thanks @ziobron for Code Review.
🏅 1 XP granted

**Review by ziobron — COMMENTED** [review]
Quick additional check

**Review by ziobron — COMMENTED** [review]
And the last check

**Comment by ziobron** [comment]
Thanks @ziobron for Code Review.
🏅 1 XP granted

File: shm/Game.cpp [hunk 48]
```diff
@@ -86,6 +84,8 @@ void Game::travel()
     else
     {
         std::cout << "There is no such island on the map! Choose another destination!" << '\n';
+        std::cin.clear();
+        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:101]: This is repeated few times, maybe create function `clearInputStream()`? 
DRY rule -> don't repeat yourself :)

File: shm/alcohol.cpp [hunk 49]
```diff
@@ -42,4 +42,8 @@ bool Alcohol::operator==(const Cargo& other) const {
 void Alcohol::nextDay(size_t elapsedTime)
 {
     basePrice_ += (elapsedTime * 2);
+}
+Cargo* Alcohol::clone()
+{
+    return new Alcohol(*this);
```
Comment (nauka-programowania-MA) [inline shm/alcohol.cpp:44]: Avoid raw ptr! and new operator

return std::unique_ptr<Cargo> and use std::make_unique.

File: shm/fruit.cpp [hunk 50]
```diff
@@ -14,7 +13,8 @@ Fruit::~Fruit()
 
 size_t Fruit::getPrice() const
 {
-    return static_cast<size_t>(basePrice_ * ((size_t)(expiry_date_ - time_elapsed_)) / expiry_date_);
+    //return static_cast<size_t>(basePrice_ * ((size_t)(expiry_date_ - time_elapsed_)) / expiry_date_);
```
Comment (nauka-programowania-MA) [inline shm/fruit.cpp]: remove commented code. 
This commented code looks good, why do you return basePrice_ instead?

File: shm/fruit.cpp [hunk 51]
```diff
@@ -36,21 +36,29 @@ Cargo &Fruit::operator-=(size_t amount)
...
     time_elapsed_ += elapsedTime;
     if (time_elapsed_ >= expiry_date_)
     {
         basePrice_ = 0;
     }
+}
+
+Cargo *Fruit::clone()
+{
+    return new Fruit(*this);
```
Comment (nauka-programowania-MA) [inline shm/fruit.cpp:58]: unique_ptr

File: shm/item.cpp [hunk 52]
```diff
@@ -45,4 +45,8 @@ bool Item::operator==(const Cargo& other) const {
 void Item::nextDay(size_t elapsedTime) 
 {
     basePrice_ += elapsedTime;    
+}
+Cargo* Item::clone()
+{
+    return new Item(*this);
```
Comment (nauka-programowania-MA) [inline shm/item.cpp]: unique_ptr

File: shm/ship.cpp [hunk 53]
```diff
@@ -53,22 +51,22 @@ Ship &Ship::operator+=(size_t num)
 void Ship::addCargo(Cargo * item)
 {
     auto cargoPtr = findMatchCargo(item);
-    if(findMatchCargo(item))
+    if(cargoPtr)
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: Better
```
if (cargoPtr = findMatchCargo(item)) {

}
```
If automatically check if this is not nullptr

File: shm/ship.cpp [hunk 54]
```diff
@@ -93,20 +91,19 @@ Cargo * Ship::findMatchCargo(Cargo * cargo)
...
-
+        // if(Fruit* f = dynamic_cast<Fruit*>(loadCargo)){
+        //     toAdd = new Fruit(*f);
+        // }
+        // if(Item* i = dynamic_cast<Item*>(loadCargo)){
+        //     toAdd = new Item(*i);
+        // }
+        // if(Alcohol* a = dynamic_cast<Alcohol*>(loadCargo)){
+        //     toAdd = new Alcohol(*a);
+        // }
```
Comment (nauka-programowania-MA) [inline shm/ship.cpp]: Remove commented code

**Review by nauka-programowania-MA — APPROVED** [review]

File: shm/Map.cpp [hunk 55]
```diff
@@ -4,57 +4,68 @@
...
     std::vector<Coordinates> cords;
     cords.reserve(constVariables::ISLANDS_COUNT);
-    for (int i = 0; i < constVariables::ISLANDS_COUNT; i++)
-    {   
-        Coordinates c (static_cast<float>(distrib(gen)), static_cast<float>(distrib(gen)));
-
+    for (int i = 0; i < (int)constVariables::ISLANDS_COUNT; i++)
+    {
+        int x = (int)distrib(gen);
+        int y = (int)distrib(gen);
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:36]: we should avoid old (C) style casting. Because it unsafe

File: shm/Map.cpp [hunk 56]
```diff
@@ -4,57 +4,68 @@
...
-    std::uniform_real_distribution<> distrib(constVariables::COORDINATE_MIN , constVariables::COORDINATE_MAX);
-
+    std::uniform_real_distribution<> distrib(constVariables::COORDINATE_MIN, constVariables::COORDINATE_MAX);
     std::vector<Coordinates> cords;
     cords.reserve(constVariables::ISLANDS_COUNT);
-    for (int i = 0; i < constVariables::ISLANDS_COUNT; i++)
-    {   
-        Coordinates c (static_cast<float>(distrib(gen)), static_cast<float>(distrib(gen)));
-
+    for (int i = 0; i < (int)constVariables::ISLANDS_COUNT; i++)
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:33]: better size_t i = 0 and don't cast to int

File: shm/Map.cpp [hunk 57]
```diff
@@ -4,57 +4,68 @@
...
+    {
+        int x = (int)distrib(gen);
+        int y = (int)distrib(gen);
+        Coordinates c(x, y);
         while (contains(cords, c))
         {
-            c.SetPositionX(distrib(gen));
-            c.SetPositionY(distrib(gen));
+            x = (int)distrib(gen);
+            y = (int)distrib(gen);
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:41]: avoid old (C) style casting

**Review by nauka-programowania-MA — APPROVED** [review]

**Pull request closed without merging**